File: apps/backend/internal/app/server/model/actress.go
package model

import (
	"encoding/json"
	"media-lib/internal/app/server/types"
	"time"

	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type Actress struct {
	ID        uint           `json:"id"`
	CreatedAt *time.Time     `json:"created_at"`
	UpdatedAt *time.Time     `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	UniqueName   string          `gorm:"type:varchar(128);uniqueIndex;not null" json:"unique_name"`
	ChineseName  string          `gorm:"type:varchar(128);uniqueIndex;not null" json:"chinese_name"`
	EnglishName  *string         `gorm:"type:varchar(128)" json:"english_name"`
	OtherNames   *datatypes.JSON `json:"other_names"`
	BirthDate    *time.Time      `json:"birth_date"`
	BirthPlace   *string         `gorm:"size:255" json:"birth_place"`
	Height       *int            `json:"height"`
	Weight       *int            `json:"weight"`
	Measurements *datatypes.JSON `json:"measurements"`
	Cup          *string         `gorm:"size:2" json:"cup"`
	BloodGroup   *string         `gorm:"size:2" json:"blood_group"`
	DebutDate    *time.Time      `json:"debut_date"`
	Hobbies      *string         `gorm:"size:255" json:"hobbies"`
	Notes        *string         `json:"notes"`
	Videos       []*Video        `gorm:"many2many:actress_video" json:"videos"`
}

func (a *Actress) DTO() *types.ActressDTO {
	if a == nil {
		return nil
	}
	return &types.ActressDTO{
		ID:           a.ID,
		UniqueName:   a.UniqueName,
		ChineseName:  a.ChineseName,
		EnglishName:  a.EnglishName,
		OtherNames:   (*json.RawMessage)(a.OtherNames),
		BirthDate:    a.BirthDate,
		BirthPlace:   a.BirthPlace,
		Height:       a.Height,
		Weight:       a.Weight,
		Measurements: (*json.RawMessage)(a.Measurements),
		Cup:          a.Cup,
		BloodGroup:   a.BloodGroup,
		DebutDate:    a.DebutDate,
		Hobbies:      a.Hobbies,
		Notes:        a.Notes,
		CreatedAt:    a.CreatedAt,
		UpdatedAt:    a.UpdatedAt,
	}
}

File: packages/service/lib/services/actress.ts
import { AxiosInstance } from 'axios';
import { ActressDTO, SubmitActressDTO, VideoDTO } from '../dtos';
import { Actress } from '../interfaces';
import { VideoService } from './video';

export class ActressService {
  #axios: AxiosInstance;

  constructor(a: AxiosInstance) {
    this.#axios = a;
  }

  static converter(dto: ActressDTO): Actress {
    return {
      id: dto.id,
      unique_name: dto.unique_name,
      chinese_name: dto.chinese_name,
      english_name: dto.english_name ?? undefined,
      other_names: dto.other_names ?? undefined,
      birth_date: dto.birth_date ?? undefined,
      birth_place: dto.birth_place ?? undefined,
      height: dto.height ?? undefined,
      weight: dto.weight ?? undefined,
      measurements: dto.measurements ?? undefined,
      cup: dto.cup ?? undefined,
      blood_group: dto.blood_group ?? undefined,
      debut_date: dto.debut_date ?? undefined,
      hobbies: dto.hobbies ?? undefined,
      notes: dto.notes ?? undefined,
      created_at: dto.created_at,
      updated_at: dto.updated_at,
    };
  }

  list = async () => {
    const { data } = await this.#axios.get<ActressDTO[]>('/actresses');
    return data.map(ActressService.converter);
  };

  create = async (dto: SubmitActressDTO) => {
    const { data } = await this.#axios.post<ActressDTO>('/actresses', dto);
    return ActressService.converter(data);
  };

  getById = async (actressId: number) => {
    const { data } = await this.#axios.get<ActressDTO>(`/actresses/${actressId}`);
    return ActressService.converter(data);
  };

  update = async (actressId: number, dto: Partial<SubmitActressDTO>) => {
    const { data } = await this.#axios.put<ActressDTO>(`/actresses/${actressId}`, dto);
    return ActressService.converter(data);
  };

  delete = async (actressId: number) => {
    await this.#axios.delete(`/actresses/${actressId}`);
  };

  listVideos = async (actressId: number) => {
    const { data } = await this.#axios.get<VideoDTO[]>(`/actresses/${actressId}/videos`);
    return data.map(VideoService.converter);
  };
}
